Core: Add LockManager to HadoopTableOperations

CodingCat opened the pull request:
this PR address the data loss issue when there are multiple threads appending to HadoopTable within the same process.

# What we observed

when we work on the migration from Delta Lake to Iceberg, we found one of the test cases where we concurrently append data to the table with Spark failed with some data loss and/or writer job abort. 

Specifically, we found when there are , e.g. 10 threads append at the same time, we eventually only get any number of versions of table between [1, 10] (with some instrumentation, I saw that multiple threads eventually committed the same version )

# Root Cause Analysis

with some analysis, we found the root cause is on the current mechanism to check version conflict. Currently

(1)  HadoopTableOperations checks the existence of target metadata file path , if exists , then throw CommitFailedException
(2) if (1) passed, HadoopTableOperations relies on the return value of FileSystem.rename to check whether the commit eventually succeeded

there are two problems here:

(a) the above (1) and (2) are not atomic, so it is possible multiple threads have passed (1) and then entered to (2) 
(b) FileSystem.rename is not sufficient to check whether the target file exists (really implementation-depended) , as a result multiple staging metadata file are renamed to the same one, e.g. multiple threads commit to v2.metadata.json eventually 

# Proposed Fix

we move the LockManagers class originally located in iceberg-aws to iceberg-core and add the capability for HadoopTables to leverage it for concurrency control in committing. 

By default, HadoopTables will use InMemoryLockManager and be only able to handle multi threading appending. Users can pass in configurations  with Hadoop Configuration or via HadoopCatalog when creating HadoopTables to switch to other implementation like DynamoDB based LockManager to handle distributed commits

CodingCat commented:
cc: @kbendick, @jackye1995, @nastra, @aokolnychyi, @danielcweeks, @rdblue 

would you mind giving a review? thanks!

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:
```diff
@@ -353,8 +345,13 @@ int findVersion() {
+      HadoopTableOperations.releaseLock.lock();
+      if (fs.exists(dst)) {
+        throw new CommitFailedException(
+                "Version %d already exists: %s", nextVersion, dst);
```
Does this need to be on a new line?

kbendick commented on spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java:
```diff
@@ -189,6 +192,56 @@ public void testAppend() throws IOException {
+    );
+
+    Dataset<Row> df = spark.createDataFrame(records, SimpleRecord.class);
+    int threadsCount = 10;
```
Nit: Same thing here about lowering the thread count.

kbendick commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +422,33 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+        put(TableProperties.COMMIT_NUM_RETRIES, "1000");
+      }
+    }, dir.toURI().toString());
+    int threadsCount = 30;
```
Nit / Somewhat blocking: Can we consider lowering this value? CI already takes some time, and I feel like spawning 30 threads in a test in our CI environment, where each of those threads runs a spark job, might be a little heavy given the very large number of concurrent users as well as the limited testing resources we get from github.

Would 4 or 5 threads be ok for all of the tests? Spawning 30 OS threads in one single test is likely to cause a large degree of contention.

What about if we changed threadsCount to 4 or 5? Would that still be ok?

kbendick commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +422,33 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+    Arrays.stream(threads).forEach(t -> {
+      try {
+        t.join();
+      } catch (InterruptedException e) {
+        // intentionally swallow to check result later
+      }
+    });
+    tableWithHighRetries.refresh();
+    assertEquals(threadsCount, Lists.newArrayList(tableWithHighRetries.snapshots()).size());
```
Is there no danger here of having a potentially flakey test? I see that the commit number of retries is set to 1000.

But our CI environment can get very busy. I worry that having to set the retries to 1000 will cause the amount of time CI takes to increase by quite a bit. We do share test resources with the entirety of the ASF after all.

kbendick reviewed (commented):
Thanks for your interest @CodingCat!

I have some concerns about this PR in general, as we don't support concurrent writers with the Hadoop catalog. I understand this is just supposed to be still best effort and from within the same process, but I do worry this will expand out into further and further effort to emulate what one should be using a catalog with a metastore of some kind (be it JDBC, HMS, DynamoDB, etc).

If others don't mind that, I won't object further.

My other concern is the incredibly large number of threads spawned in some of the tests. I think in our busy CI environment that this is going to be a real problem for stability, if not making the tests somewhat flaky. Could we reduce the number of threads spawned in each test to 2, or even 4 or so? 30 is definitely way too many, and 10 is still quite a lot for one test. We do share the CI infrastructure with a large number of other projects in the Apache Foundation.

rdblue commented:
@CodingCat, `HadoopTableOperations` is not safe to use with a file system that doesn't support atomic rename. Even if we were to add a lock as proposed in this PR, concurrent processes could still produce the same version in stores like S3 or even most local file systems.

If you want to avoid using a catalog implementation that provides an atomic rename (like JDBC, DynamoDB, or Hive) then you'd need to use a lock that can coordinate across processes.

For using Glue as a metastore, @jackye1995 added a DynamoDB based lock, in `DynamoLockManager`. I recommend adding the ability for `HadoopTableOperations` to use the DynamoDB lock instead of adding a JVM lock.

CodingCat commented:
Hi, @kbendick and @rdblue , thank you very much for the review! IIUC, besides the too-many thread number concern (which I will address with code), your comments are both related to the scope of HadoopTable. let me try to address them here and love to hear if it makes sense to you

**First, this PR is exactly only for multi threading instead of processes.** The reason we care about multi threading operation is that it is much more implicit than multiple processes. Users may easily realize there are 1+ Spark applications running in the cluster but not necessarily realized there are multiple threads appending to the same table from the same Spark application. Just like our case which is somewhat an event-driven append case and each event is processed by a Spark job triggered from a separate thread. So as @kbendick said, it is a best-effort work and as @rdblue said, the multiple processes limitation is still there. 

Second, regarding @kbendick 's concern,  **how to prevent it extending to an endless effort to make HadoopTable reaching to the capability of other catalog implementations, i.e. make HadoopTable eventually support distributed commits** . I think it is a valid concern on the resource allocation of community and roadmap of the project . My personal opinion is to try to find some way to explicitly fail the application (in this PR or the other, depends on the workload) when finding multiple processes are committing, not only serving as a hard guard but also prevent users didn't read docs and just to lose some data

Third, regarding @rdblue's suggestion about **making HadoopTableOperations support distributed committing by leveraging DynamoLockManager** . I am hesitated to go with this path for now , please do let me know if I misunderstood anything: essentially we are going to make iceberg-core depends on iceberg-aws which will bring the circular dependency and also some counter-intuitive architecture with a hardcode component from AWS...

Lastly, as I said in the background info of the PR, companies like us have been operating in some company's env for a while. There was a period that every customer was encouraged to write parquet path instead of registering a table in HMS and it has make 100s of our data applications are talking to file system directly instead of HMS. The effort here is also about making functionality parity with Delta Lake



Regarding the action items:

1. I will definitely address thread number issue, the retry-limit like 1000 or thread count are more like conservative numbers, easy to lower down. in theory , retry-limit can be <= thread count
2. if you all agree, I will try to find some way to fail applications explicitly when there are distributed commits in HadoopTable...I will share my findings and depends on the complexity, we can decide how to move forward

Love to hear your feedbacks!

nastra commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +422,33 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+    }
+    Arrays.stream(threads).forEach(t -> {
+      try {
+        t.join();
```
this is unfortunately a recipe for flaky tests. It would be much nicer if we could use https://github.com/awaitility/awaitility and have something like `Awaitility.await().atMost(X, SECONDS).untilAsserted(() -> assertThat(...).isEqualTo(..))`

rdblue replied:
I agree in general, but in this case there are a couple of easier solutions.

First, these tests already exist for other catalogs so we can adapt them to use `HadoopCatalog. This should copy `TestJdbcTableConcurrency` and update it. Those tests aren't flaky.

Second, we have a trusted way to run tasks in a thread pool, the `Tasks` class that is used all over Iceberg. Using that is a really easy way to run tests like this. It is what `TestJdbcTableConcurrency` uses.

If that doesn't work, then a library like Awaitility works for me.

nastra commented on spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java:
```diff
@@ -189,6 +192,56 @@ public void testAppend() throws IOException {
+        .load(location.toString());
+
+    List<SimpleRecord> actual = result.orderBy("id").as(Encoders.bean(SimpleRecord.class)).collectAsList();
+    Assert.assertEquals("Number of rows should match", 3 * threadsCount, actual.size());
```
same comment here about using Awaitility to assert async code

nastra reviewed (commented):
I would be much in favor of using something like https://github.com/awaitility/awaitility to test async code as this would reduce the chance for flakiness. @rdblue thoughts?

rdblue commented:
@CodingCat, adding the ability to use a lock doesn't need to require a dependency on iceberg-aws. The lock manager was built as an API and multiple implementations, including an in-memory implementation. The interface and in-memory implementation could be moved into core and you would then dynamically load your preferred lock manager. That avoids the circular dependency and adds very similar functionality to what you're doing here out of the box.

Regarding the use case, I understand wanting to have no dependency other than the object store. I think the right way to do that is to coordinate writers, whether those are threads or processes. If you choose to go with the in-memory lock manager and risk data loss that's up to you -- to your point, you would have reached parity so you may not want full locking.

jackye1995 commented:
This is actually interesting development. In the recent announcement related to AWS Athena support for Iceberg, we mentioned that the Glue catalog integration uses Glue native optimistic locking and no longer require any lock such as DynamoLockManager. So we are planning to deprecate the lock manager once the new Glue SDK with that feature is released in a few weeks. 

We can move the LockManager interface and in-memory implementation to core if that is the place which needs this feature the most.

CodingCat commented:
@rdblue @jackye1995 @kbendick I have moved LockManagers to core and added the capability for HadoopTableOperations to leverage that. Specifically, users can pass in lock configuration via Hadoop Configuration (prefix with `iceberg.`)  , please let me know if this is the approach you like. And also I reduced the number of threads in tests

rdblue commented:
Looks like there are Javadoc/build problems, but tests are passing so we can review while you fix those in parallel.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:
```diff
@@ -63,15 +64,17 @@
     this.conf = conf;
     this.location = location;
     this.fileIO = fileIO;
+    this.lockManager = lockMgr;
```
In Iceberg, we avoid abbreviations like `manager -> mgr` because it hurts readability and isn't actually easier to type.

rdblue commented on aws/src/integration/java/org/apache/iceberg/aws/glue/TestDynamoLockManager.java:
```diff
@@ -153,9 +153,9 @@ public void testAcquireSingleProcess() throws Exception {
-        CatalogProperties.LOCK_TABLE, lockTableName
+        LockManagerProperties.LOCK_ACQUIRE_INTERVAL_MS, "500",
+        LockManagerProperties.LOCK_ACQUIRE_TIMEOUT_MS, "100000000",
+        LockManagerProperties.LOCK_TABLE, lockTableName
```
I don't think these properties need to be relocated. They are still properties that are passed to the catalog, so we should leave them in `CatalogProperties`.

CodingCat replied:
hmmm...not necessarily? users can create HadoopTables directly and then create HadoopTableOperations (assuming we agree on the current parameter passing mechanism that users pass in via Hadoop Configurations)

jackye1995 replied:
These properties cannot change class for backwards compatibility. 

And I don't think it breaks definition of `CatalogProperties` to keep these configurations there, because there are 2 places using this, (1) hadoop catalog, (2) hadoop tables. For (1), it's still in a catalog, and for (2), you are technically not using these properties directly, but only using it with a pre-defined prefix. That is a custom definition in `HadoopTables` to use these properties in the specific way, I don't think we need to also move properties to another class to accommodate this use case.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java:
```diff
@@ -105,6 +108,8 @@ public void initialize(String name, Map<String, String> properties) {
 
     this.suppressPermissionError = Boolean.parseBoolean(properties.get(HADOOP_SUPPRESS_PERMISSION_ERROR));
+
+    this.lockManager = LockManagers.from(properties);
```
Nit: Can you put this just after `fileIO` to match the order of the field declarations?

CodingCat replied:
done

jackye1995 replied:
I don't think we should do this for `HadoopCatalog`, because hadoop catalog users usually can leverage Hadoop file system's mutual exclusive writing feature and does not need a pessimistic lock to do commit. I think we should allow `LockManager` to be null in `HadoopTableOperations`, and only if it is configured, we use it. 

This probably also means we need another config property `lock-enabled`.

CodingCat replied:
@jackye1995 I think users can still use HadoopCatalog with S3 or local file system even that's not the optimal solution, so the data loss issue is still there?

rdblue replied:
@jackye1995, I don't think there's a problem using the lock manager by default, considering the implementation is in-memory only. That's not going to cause much overhead.

rdblue commented on core/src/test/java/org/apache/iceberg/util/TestInMemoryLockManager.java:
```diff
@@ -1,23 +1,18 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
Please revert the license changes.

CodingCat replied:
ah, sorry, seems intelliJ added this somehow

rdblue commented on core/src/main/java/org/apache/iceberg/util/LockManagerProperties.java:
```diff
@@ -0,0 +1,42 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
I don't think we need this file.

CodingCat replied:
as said above, with HadoopTables, users can directly create a HadoopTables object by passing in lock configurations, so it is not exclusively a catalog config anymore?

rdblue commented on core/src/test/java/org/apache/iceberg/util/TestLockManagers.java:
```diff
@@ -1,26 +1,20 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
Wrong license text here as well. No need for this to change.

CodingCat replied:
done

rdblue commented on spark/v2.4/spark/src/test/java/org/apache/iceberg/spark/source/TestSparkDataWrite.java:
```diff
@@ -189,6 +192,56 @@ public void testAppend() throws IOException {
   }
 
+  @Test
+  public void testConcurrentAppend() throws IOException {
```
I don't think that we need tests for individual engines or versions. The catalog should be thoroughly tested and its behavior won't change when it is embedded in an engine.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +203,24 @@ TableOperations newTableOps(String location) {
+  }
+
+  private LockManager createOrGetLockManager(String location) {
+    Map<String, String> properties = new HashMap<>();
```
Nit: use `Maps.newHashMap()` instead.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +203,24 @@ TableOperations newTableOps(String location) {
+    }
+  }
+
+  private LockManager createOrGetLockManager(String location) {
```
Nit: most APIs use `getOrCreate` instead of `createOrGet`.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +203,24 @@ TableOperations newTableOps(String location) {
+
+  private LockManager createOrGetLockManager(String location) {
+    Map<String, String> properties = new HashMap<>();
+    Iterator<Map.Entry<String, String>> configEntries = this.conf.iterator();
```
Nit: no need for the prefix `this.` when accessing a field. We use it when setting a field so that it is easy to see that state is modified, but we don't use it on read.

CodingCat replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +203,24 @@ TableOperations newTableOps(String location) {
     } else {
-      return new HadoopTableOperations(new Path(location), new HadoopFileIO(conf), conf);
+      return new HadoopTableOperations(new Path(location), new HadoopFileIO(conf), conf,
+              createOrGetLockManager(location));
```
This class should only need one `LockManager`. The same `LockManager` is thread safe and can be shared by all the `TableOperations`. That's how the original use in `GlueCatalog` works.

CodingCat replied:
hmmm...the consideration here is actually users can pass different configurations of LockManager for different tables

rdblue replied:
The `LockManager` is configured by the Hadoop `Configuration`, which is passed into `HadoopTables` at create time. While this does implement `Configurable`, I don't think there is any need to support changing lock manager configurations.

If we tried to support configuration changes we would introduce inconsistency: what happens when a table is loaded, the config changes, and then the table is loaded again? Either you use the same lock manager with the old settings, or you create a different lock manager and break in-memory locks. There is no need to introduce these worries.

To avoid unnecessary complexity, I think we should remove the cache and instantiate a lock manager in the constructor that is used for all tables loaded by this interface. Then we don't have to worry about having a lock manager per table -- that's a good thing because each lock manager has a heartbeat thread!

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:
```diff
@@ -353,8 +346,12 @@ int findVersion() {
-  private void renameToFinal(FileSystem fs, Path src, Path dst) {
+  private void renameToFinal(FileSystem fs, Path src, Path dst, int nextVersion) {
     try {
+      lockManager.acquire(dst.toString(), Thread.currentThread().getName());
```
The current thread name can repeat across processes. This should be something unique. The `GlueCatalog` uses the metadata file to commit. I think this should do the same and use `src`.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:
```diff
@@ -353,8 +346,12 @@ int findVersion() {
+      lockManager.acquire(dst.toString(), Thread.currentThread().getName());
+      if (fs.exists(dst)) {
+        throw new CommitFailedException("Version %d already exists: %s", nextVersion, dst);
+      }
```
Nit: please add a newline between control flow blocks and the following statement.

CodingCat replied:
done

CodingCat commented:
Hi, all,

Thanks for the review!

I am currently on vacation without my laptop , will address the comments
after Dec 20

Thanks!

Nan

On Sun, Dec 12, 2021 at 9:25 AM Ryan Blue ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java
> <https://github.com/apache/iceberg/pull/3663#discussion_r767316693>:
>
> >      try {
> +      lockManager.acquire(dst.toString(), Thread.currentThread().getName());
> +      if (fs.exists(dst)) {
> +        throw new CommitFailedException("Version %d already exists: %s", nextVersion, dst);
> +      }
>
> Nit: please add a newline between control flow blocks and the following
> statement.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/iceberg/pull/3663#pullrequestreview-829660825>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFFQ6F425IPL6VFFVROI63UQTZJRANCNFSM5JKZG27A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

jackye1995 commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +203,24 @@ TableOperations newTableOps(String location) {
+    while (configEntries.hasNext()) {
+      Map.Entry<String, String> entry = configEntries.next();
+      String key = entry.getKey();
+      if (key.startsWith("iceberg.lock.")) {
```
"iceberg." should be a public static variable because a process should import this variable to construct lock properties. And I wonder if we should use something more complex to avoid namespace collision, such as "iceberg.tables.hadoop."

CodingCat replied:
done

jackye1995 commented on core/src/main/java/org/apache/iceberg/util/LockManager.java:
```diff
@@ -17,14 +17,14 @@
  */
 
-package org.apache.iceberg.aws.glue;
+package org.apache.iceberg.util;
```
why is this in the `util` package? If we want to properly support it, it feels more nature to me to add this to `iceberg-api`.

CodingCat replied:
done

jackye1995 commented:
Regarding the `@apiNote` tag issue, found this:

> These tags are enabled by use of the -tag feature on the javadoc tool command line. They are not proposed as standard javadoc tags and may be implemented differently in future Java releases.

Since this is not standard, let's just remove `@apiNote` tag and use plain sentence instead.

rdblue commented on api/src/main/java/org/apache/iceberg/LockManager.java:
```diff
@@ -17,14 +17,14 @@
  */
 
-package org.apache.iceberg.aws.glue;
+package org.apache.iceberg;
```
@jackye1995, couldn't this still be in the `util` package even if it is in the API module?

jackye1995 replied:
yes it can still be in the `util` package. 

I think the actual module and package of this class depends on how we want to position this feature. I remember you mentioned we might want to eventually introduce some kind of pessimistic lock to fully lock a table to make certain operations succeed.

So the following makes sense to me:
1. if it's only for `HadoopTables` (or maybe `HadoopCatalog`), then `iceberg-core` module can work and it can either be in `org.apache.iceberg.hadoop` or `org.apache.iceberg.util`.
2. if we want this to do other things in the future like fully locking a table in a catalog, then I think `iceberg-api` module and `org.apache.iceberg` package seems more official.

rdblue replied:
I think that having this in the API module makes sense, but I would put `LockManagers` and the in-memory implementation in core. I'd also move the implementation back to the `util` package since we want to avoid making `org.apache.iceberg` any more huge.

rdblue commented on core/src/main/java/org/apache/iceberg/CatalogProperties.java:
```diff
@@ -36,6 +36,9 @@ private CatalogProperties() {
   public static final long CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS_DEFAULT = TimeUnit.MINUTES.toMillis(5);
 
+  public static final String APP_ID = "app-id";
+  public static final String USER = "user";
```
Can you revert the changes to this file, please? This doesn't look like a necessary change.

CodingCat replied:
hmmm...strange, seems unsynced from master branch

rdblue commented on core/src/main/java/org/apache/iceberg/CatalogProperties.java:
```diff
@@ -57,5 +57,4 @@ private CatalogProperties() {
 
   public static final String APP_ID = "app-id";
   public static final String USER = "user";
-
```
Looks like this should be reverted as well.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopCatalog.java:
```diff
@@ -104,6 +107,8 @@ public void initialize(String name, Map<String, String> properties) {
     String fileIOImpl = properties.get(CatalogProperties.FILE_IO_IMPL);
     this.fileIO = fileIOImpl == null ? new HadoopFileIO(conf) : CatalogUtil.loadFileIO(fileIOImpl, properties, conf);
 
+    this.lockManager = LockManagers.from(properties);
```
@CodingCat, the lock manager API has a `close` method that I think should be called when the catalog is closed. Can you implement closing similar to what's in the `GlueCatalog`?

CodingCat replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:
```diff
@@ -353,8 +346,13 @@ int findVersion() {
-  private void renameToFinal(FileSystem fs, Path src, Path dst) {
+  private void renameToFinal(FileSystem fs, Path src, Path dst, int nextVersion) {
     try {
+      lockManager.acquire(dst.toString(), src.toString());
```
For future me: this looks good.

rdblue commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +424,29 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+    int threadsCount = 5;
+    Table tableWithHighRetries = TABLES.create(SCHEMA, SPEC, new HashMap<String, String>() {
+      {
+        put(TableProperties.COMMIT_NUM_RETRIES, String.valueOf(threadsCount));
```
Instead of `new HashMap`, we prefer `Maps.newHashMap` in most cases. But here, you can use `ImmutableMap` to more succinctly create a map: `ImmutableMap.of(COMMIT_NUM_RETRIES, String.valueOf(threadsCount))`.

CodingCat replied:
done

rdblue commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +424,29 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
   }
+
+  @Test
+  public void testConcurrentAppend() throws Exception {
```
I's fine to add more tests, but please copy the `testConcurrentFastAppends` test from JDBC into this suite. We have validated that it catches quite a few problems so we should reuse it.

CodingCat replied:
done

rdblue replied:
Can you rename the method to match the source?

rdblue replied:
Actually, can you copy the original without modification other than to set up the catalog? It is too heavily modified from the original version.

CodingCat replied:
actually there are no significant changes, just making some constant in variable ....and changing the values of those constant, 

the major reason to change to larger values is to make the issue easier to reproduce (it is hard to see the failure of committing with only 2 threads as set in JDBC test)

rdblue commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +429,45 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+  @Test
+  public void testConcurrentAppend() throws Exception {
+    assertTrue("Should create v1 metadata",
+            version(1).exists() && version(1).isFile());
```
Is this newline necessary?

rdblue commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +429,45 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+            barrier.incrementAndGet();
+          }
+        });
+    tableWithHighRetries.refresh();
```
Can you add newlines to match the test that this was copied from? There are no newlines and we typically like to break up the code more than this.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +204,25 @@ TableOperations newTableOps(String location) {
+      }
+      lockManager = LockManagers.from(properties);
     }
+    return lockManager;
```
Nit: newline between control flow and the following statement.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -54,8 +58,14 @@
 public class HadoopTables implements Tables, Configurable {
+
+  public static final String LOCK_PROPERTY_PREFIX = "iceberg.tables.hadoop.";
+
```
Can you make this private and remove the newline? I don't think there's a reason for this prefix to be public.

CodingCat replied:
this is actually to address the comments at https://github.com/apache/iceberg/pull/3663#discussion_r774119018 from  @jackye1995

rdblue replied:
That's okay, then.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -194,8 +204,25 @@ TableOperations newTableOps(String location) {
+          properties.put(key.substring(LOCK_PROPERTY_PREFIX.length()), entry.getValue());
+        }
+      }
+      lockManager = LockManagers.from(properties);
```
Nit: newline between control flow and the following statement.

rdblue commented:
Thanks, @CodingCat! I merged this. It should work as you intend now, although I still recommend using a real lock manager in production!

rdblue commented on core/src/test/java/org/apache/iceberg/hadoop/TestHadoopCommits.java:
```diff
@@ -419,4 +429,47 @@ public void testCanReadOldCompressedManifestFiles() throws Exception {
+
+    tableWithHighRetries.refresh();
+    assertEquals(threadsCount * numberOfCommitedFilesPerThread,
+        Lists.newArrayList(tableWithHighRetries.snapshots()).size());
```
I ran this locally without the lock manager additions and it failed as expected.

yihaizhu replied:
I constantly run with this test failure locally as:
org.apache.iceberg.hadoop.TestHadoopCommits > testConcurrentFastAppends FAILED
    java.lang.AssertionError: expected:<50> but was:<40>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.apache.iceberg.hadoop.TestHadoopCommits.testConcurrentFastAppends(TestHadoopCommits.java:472)

1353 tests completed, 1 failed, 21 skipped

Is this expected?

yihaizhu replied:
I run it again and this time i get the actual value: 45. The actual value seems to change at every run.

org.apache.iceberg.hadoop.TestHadoopCommits > testConcurrentFastAppends FAILED
    java.lang.AssertionError: expected:<50> but was:<45>
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.failNotEquals(Assert.java:835)
        at org.junit.Assert.assertEquals(Assert.java:647)
        at org.junit.Assert.assertEquals(Assert.java:633)
        at org.apache.iceberg.hadoop.TestHadoopCommits.testConcurrentFastAppends(TestHadoopCommits.java:472)

BsoBird commented:
@CodingCat 
Hello, I am working on a fix related to HadoopTableOperations. I would like to ask you a question.
```
      if (!lockManager.acquire(dst.toString(), src.toString())) 
```
Why are you using dst.toString() and src.toString() as lock-key.
